src/GafferScene/Instancer.cpp

```diff
@@ -2401,3 +2568,246 @@ void Instancer::PrototypeScope::setPrototype( const EngineData *engine, const Sc
+	Box3f bound = prototypesPlug->boundPlug()->getValue();
+
+	// const_cast should be safe since the elements of the attributes vector are never modified
+	attributes.push_back( const_cast< CompoundObject*>( prototypesPlug->attributesPlug()->getValue().get() ) );
+
+	for( unsigned int i = 0; i < sampleTimes.size(); i++ )
+	{
+		scope.setFrame( sampleTimes[i] );
+		transforms.push_back( prototypesPlug->transformPlug()->getValue() );
+	}
+
+	scope.remove( ScenePlug::scenePathContextName );
+
+	IECore::MurmurHash h = hash;
+	h.append( index );
+
+	// We find the capsules using the engine at shutter open, but the time used to construct the capsules
+	// must be the on-frame time, since the capsules will add their own shutter
+	scope.setFrame( onFrameTime );
+
+
+	CapsulePtr newCapsule = new Capsule(
+		prototypesPlug,
+		*prototypeRoot,
+		*scope.context(),
+		h,
+		bound
+	);
+
+	// Pass through our render options to the sub-capsules
+	newCapsule->setRenderOptions( renderOptions );
+
+	objects.push_back( std::move( newCapsule ) );
+
+}
+
+// \todo - this can be omitted once we add a specialization of std::hash for MurmurHash to Cortex
+struct MurmurHashHasher
+{
+	std::size_t operator()(const IECore::MurmurHash& h) const
+	{
+		return h.h1();
+	}
+};
+
+} // namespace
+
+void Instancer::InstancerCapsule::render( IECoreScenePreview::Renderer *renderer ) const
+{
+	if( !renderer )
+	{
+		throw IECore::Exception( "Null renderer passed to InstancerCapsule" );
+	}
+	throwIfNoScene();
+
+	float onFrameTime = context()->getFrame();
+	Context::EditableScope scope( context() );
+
+	ScenePlug::ScenePath enginePath( root().begin(), root().begin() + root().size() - 2 );
+
+	IECore::ConstCompoundObjectPtr globals = m_instancer->inPlug()->globals();
+	ConstCompoundObjectPtr sceneAttributes = m_instancer->inPlug()->fullAttributes( enginePath );
+
+	FloatVectorDataPtr sampleTimesData = new FloatVectorData;
+	vector<float> &sampleTimes = sampleTimesData->writable();
+
+	std::optional<GafferScene::Private::RendererAlgo::RenderOptions> renderOptions = getRenderOptions();
+	if( !renderOptions )
+	{
+		renderOptions = GafferScene::Private::RendererAlgo::RenderOptions( scene() );
+	}
+	GafferScene::Private::RendererAlgo::transformMotionTimes( *renderOptions, sceneAttributes.get(), sampleTimes );
+
+	if( sampleTimes.size() == 0 )
+	{
+		sampleTimes.push_back( context()->getFrame() );
+	}
+
+	std::vector< ConstEngineDataPtr > engines( sampleTimes.size() );
+	for( unsigned int i = 0; i < sampleTimes.size(); i++ )
+	{
+		scope.setFrame( sampleTimes[i] );
+		engines[i] = m_instancer->engine( enginePath, scope.context() );
+	}
+
+	scope.setFrame( onFrameTime );
+
+	const std::vector<int> &instancesForPrototype = engines[0]->instancesForPrototype( root().back() );
+	const ScenePlug::ScenePath *prototypeRoot = engines[0]->prototypeRoot( root().back() );
+
+	const ScenePlug *prototypesPlug = m_instancer->prototypesPlug();
+
+	ObjectVectorPtr prototypesData = new ObjectVector();
```

Comment:
Yeah, good point. This current structure is somewhat leftover from when I needed to cram all this into an Object to pass to the renderer ... it will be a bit annoying to restructure things again if we have to go back to that, but at this point, it seems like we'll cross that bridge when we come to it.